Assets/Mirror/Editor/Weaver/WeaverExceptions.cs

```diff
@@ -0,0 +1,25 @@
+using System;
+using Mono.CecilX;
+
+namespace Mirror.Weaver
+{
+    [Serializable]
+    public abstract class WeaverException : Exception
```

Comment:
Do we really need custom exception types?
The way I see it, weaver should fail if any exception happens, since the result is not usable no matter what.
So we might as well just throw new Exception("message") and catch them nowhere?

Reply:
If there are 10 mistakes with the users code we want to show 10 error messages, so we don't just want to stop everything after 1 error. For example we want to process all NetworkBehaviours, even if one of them has an error. The end result will still be unusable but it would show all mistakes allowing the user to fix them all at once.

Custom exception type allows us to include the MemberReference making it easier to throw errors. We dont want to have to worry about adding `(at ...)`  to every message before we throw it.

Custom exceptions also help us keep track of what functions throws what exception, for example the writers will throw `GenerateWriterException` but SyncVars will throw `SyncVarException`

Having multiple exception means that we dont have to deal with writing an error message where we throw them, we would just need to throw the type of error and why it happens, then in a different part of the code we worry about crafting the error message into something that the user can understand.

Here is an example of how we create the error message inside the exception, meaning that the place we throw it just needs to worry about passing in the arguments. 
https://github.com/vis2k/Mirror/blob/ec4ae5122040155ff8d30c55b2fb6d4a8a147c65/Assets/Mirror/Editor/Weaver/WeaverExceptions.cs#L31-L35
https://github.com/vis2k/Mirror/blob/ec4ae5122040155ff8d30c55b2fb6d4a8a147c65/Assets/Mirror/Editor/Weaver/Processors/NetworkBehaviourProcessor.cs#L434-L437

Reply:
We are basically replacing code like this
```cs
                MethodReference writeFunc = Writers.GetWriteFunc(syncVar.FieldType);
                if (writeFunc != null)
                {
                    worker.Append(worker.Create(OpCodes.Call, writeFunc));
                }
                else
                {
                    Weaver.Error($"{syncVar.Name} has unsupported type. Use a supported Mirror type instead", syncVar);
                    return;
                }
```
With code like this
```cs
                try
                {
                    MethodReference writeFunc = Writers.GetWriteFunc(syncVar.FieldType);
                    worker.Append(worker.Create(OpCodes.Call, writeFunc));
                }
                catch (WeaverException e)
                {
                    throw new SyncVarException(syncVar, e);
                }
```

In the 1st we have to worry about if things are null or not, and what we want to show to the user. 
In the 2nd we just run a method, and it either works or throw. If it does throw then we throw it up high with the new Exception type.